File: 백준/Silver/3085. 사탕 게임/사탕 게임.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	static int n;
	static char[][] data;
	static int answer = 0;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());
		data = new char[n][n];
		
		String line="";
		for (int i = 0; i < n; i++) {
			line = br.readLine();
			data[i] = line.toCharArray();
		}
		
		for(int i = 0; i < n; i++) {
			for (int j = 0; j < n-1; j++) {
				char temp = data[i][j]; 
				data[i][j] = data[i][j+1];
				data[i][j+1] = temp;
				answer = Math.max(check(), answer);
				temp = data[i][j]; 
				data[i][j] = data[i][j+1];
				data[i][j+1] = temp;
			}
		}
		for(int i = 0; i < n; i++) {
			for (int j = 0; j < n-1; j++) {
				char temp = data[j][i]; 
				data[j][i] = data[j+1][i];
				data[j+1][i] = temp;
				answer = Math.max(check(), answer);
				temp = data[j][i]; 
				data[j][i] = data[j+1][i];
				data[j+1][i] = temp;
			}
		}
		System.out.println(answer);
		
	}
	
	private static int check() {
		for(int i = 0; i < n; i++) {
            int cnt = 1;
            for(int j = 0; j < n-1; j++) {
                if(data[i][j] == data[i][j+1]) {
                    cnt++;
                } else {
                    cnt = 1;
                }
                answer = Math.max(answer, cnt);
            }
        }

        for(int i = 0; i < n; i++) {
            int cnt = 1;
            for(int j = 0; j < n-1; j++) {
                if(data[j][i] == data[j+1][i]) {
                    cnt++;
                } else {
                    cnt = 1;
                }
                answer = Math.max(answer, cnt);
            }
        }

        return answer;
	}
	


}

File: 백준/Silver/3085. 사탕 게임/사탕 게임.py
import sys
Input = sys.stdin.readline

def count_box(box):
    Max_result = 1

    for i in range(a):

        #열 순회
        count = 1
        for j in range(1,a):
            if(box[i][j] == box[i][j-1]): # 왼쪽, 오른쪽 비교
                count += 1 # 이전 것과 같다면 cnt에 1 더하기
            else:
                count = 1 # 이전과 다르다면 다시 1로 초기화

            if(count > Max_result ):
                Max_result = count

        count = 1        
        for j in range(1, a):
            if(box[j][i] == box[j-1][i]): # 위아래 비교
                count += 1
            else:
                count = 1
            if(count > Max_result ):
                Max_result = count

    return Max_result

a = int(Input())
box = [list(Input()) for _ in range(a) ]
answer = 0

for i in range(a):
    for j in range(a):
        #열 바꾸기(가로)
        if(j+1 < a):
            #인접한것 바꾸기, 같은것도 바꾼다.
            box[i][j], box[i][j+1] =box[i][j+1], box[i][j]
            temp = count_box(box)

            answer = max(temp, answer)

            #다시 돌려놓기
            box[i][j], box[i][j+1] =box[i][j+1], box[i][j]

        #행바꾸기(세로)
        if(i+1 < a):
            #인접한것 바꾸기, 같은것도 바꾼다.
            box[i][j], box[i+1][j] =box[i+1][j], box[i][j]
            temp = count_box(box)
            answer = max(temp, answer)

            #다시 돌려놓기
            box[i][j], box[i+1][j] =box[i+1][j], box[i][j]

print(answer)